pkg/composableschemadsl/compiler/compiler.go

```diff
@@ -86,55 +86,39 @@ type Option func(*config)
 
 type ObjectPrefixOption func(*config)
 
-type compilationContext struct {
-	// The set of definition names that we've seen as we compile.
-	// If these collide we throw an error.
-	existingNames *mapz.Set[string]
-	// The global set of files we've visited in the import process.
-	// If these collide we short circuit, preventing duplicate imports.
-	globallyVisitedFiles *mapz.Set[string]
-	// The set of files that we've visited on a particular leg of the recursion.
-	// This allows for detection of circular imports.
-	// NOTE: This depends on an assumption that a depth-first search will always
-	// find a cycle, even if we're otherwise marking globally visited nodes.
-	locallyVisitedFiles *mapz.Set[string]
-}
-
 // Compile compilers the input schema into a set of namespace definition protos.
 func Compile(schema InputSchema, prefix ObjectPrefixOption, opts ...Option) (*CompiledSchema, error) {
-	cctx := compilationContext{
-		existingNames:        mapz.NewSet[string](),
-		globallyVisitedFiles: mapz.NewSet[string](),
-		locallyVisitedFiles:  mapz.NewSet[string](),
-	}
-	return compileImpl(schema, cctx, prefix, opts...)
-}
-
-func compileImpl(schema InputSchema, cctx compilationContext, prefix ObjectPrefixOption, opts ...Option) (*CompiledSchema, error) {
 	cfg := &config{}
 	prefix(cfg) // required option
 
 	for _, fn := range opts {
 		fn(cfg)
 	}
 
-	mapper := newPositionMapper(schema)
-	root := parser.Parse(createAstNode, schema.Source, schema.SchemaString).(*dslNode)
-	errs := root.FindAll(dslshape.NodeTypeError)
-	if len(errs) > 0 {
-		err := errorNodeToError(errs[0], mapper)
+	root, mapper, err := parseSchema(schema)
+	if err != nil {
 		return nil, err
 	}
 
-	compiled, err := translate(translationContext{
-		objectTypePrefix:     cfg.objectTypePrefix,
-		mapper:               mapper,
-		schemaString:         schema.SchemaString,
-		skipValidate:         cfg.skipValidation,
+	// NOTE: import translation is done separately so that partial references
+	// and definitions defined in separate files can correctly resolve.
+	err = translateImports(importTranslationContext{
```

Comment:
Note that the import process doesn't produce a new tree - it operates on the tree started in the root file and recursively and mutatively flattens things on. I'm not entirely sure I like the approach; I'd welcome other suggestions.